network/page.go

```diff
@@ -0,0 +1,220 @@
+// graph g using the given damping factor and terminating when the 2-norm of the
+// vector difference between iterations is below tol. The returned map is
+// keyed on the graph node IDs.
+func PageRankSparse(g graph.DirectedGraph, damp, tol float64) map[int]float64 {
+	// PageRankSparse is implemented according to "How Google Finds Your Needle
+	// in the Web's Haystack".
+	//
+	// G.I^k = alpha.H.I^k + alpha.A.I^k + (1-alpha).1/n.1.I^k
+	//
+	// http://www.ams.org/samplings/feature-column/fcarc-pagerank
+
+	nodes := g.NodeList()
+	indexOf := make(map[int]int, len(nodes))
+	for i, n := range nodes {
+		indexOf[n.ID()] = i
+	}
+
+	m := make(rowCompressedMatrix, len(nodes))
+	var dangling compressedRow
+	df := damp / float64(len(nodes))
+	for j, u := range nodes {
+		to := g.Successors(u)
+		f := damp / float64(len(to))
+		for _, v := range to {
+			m.addTo(indexOf[v.ID()], j, f)
+		}
+		if len(to) == 0 {
+			dangling.addTo(j, df)
+		}
+	}
+
+	last := make([]float64, len(nodes))
+	for i := range last {
+		last[i] = 1
+	}
+	lastV := mat64.NewVector(len(nodes), last)
+
+	vec := make([]float64, len(nodes))
+	var sum float64
+	for i := range vec {
+		r := rand.NormFloat64()
+		sum += r
+		vec[i] = r
+	}
+	f := 1 / sum
+	for i := range vec {
+		vec[i] *= f
+	}
+	v := mat64.NewVector(len(nodes), vec)
+
+	dt := (1 - damp) / float64(len(nodes))
+	for {
+		lastV, v = v, lastV
+
+		m.mulVecUnitary(v, lastV)          // First term of the G matrix equation;
+		with := dangling.dotUnitary(lastV) // Second term;
+		away := onesDotUnitary(dt, lastV)  // Last term.
+
+		floats.AddConst(with+away, v.RawVector().Data)
+		if normDiff(vec, last) < tol {
+			break
+		}
+	}
+
+	ranks := make(map[int]float64, len(nodes))
+	for i, r := range v.RawVector().Data {
+		ranks[nodes[i].ID()] = r
+	}
+
+	return ranks
+}
+
+// rowCompressedMatrix implements row-compressed
+// matrix/vector multiplication.
+type rowCompressedMatrix []compressedRow
+
+// addTo adds the value v to the matrix element at (i,j).
+func (m rowCompressedMatrix) addTo(i, j int, v float64) { m[i].addTo(j, v) }
+
+// mulVecUnitary multiplies the receiver by the source vector.
+// It assumes src and dst are the same length as m.
+func (m rowCompressedMatrix) mulVecUnitary(dst, src *mat64.Vector) {
+	dMat := dst.RawVector().Data
+	for i, r := range m {
+		dMat[i] = r.dotUnitary(src)
+	}
+}
+
+// compressedRow implements a simplified scatter-based Ddot.
+type compressedRow []sparseElement
+
+// addTo adds the value v to the vector element at j.
+func (r *compressedRow) addTo(j int, v float64) { *r = append(*r, sparseElement{index: j, value: v}) }
+
+// dotUnitary performs a simplified scatter-based Ddot
+// operations on v and the receiver.
+func (r compressedRow) dotUnitary(v *mat64.Vector) float64 {
+	var sum float64
+	vec := v.RawVector().Data
+	for _, e := range r {
+		sum += vec[e.index] * e.value
+	}
+	return sum
+}
+
+// sparseElement is a sparse matrix element.
```

Comment:
To explain further, there are two major categories of sparse element, only triple elements know anything about more than a linear dimension, and those are reasonably rarely used in practice. When the sparse package is implemented, a type like this will be a matrix and vector element.